sw/device/tests/power_virus_systemtest.c

```diff
@@ -451,7 +480,114 @@ static void configure_spi_host() {
+
+  // Load data into HMAC block.
+  hmac_testutils_push_message(&hmac, kHmacRefData, sizeof(kHmacRefData));
+  hmac_testutils_check_message_length(&hmac, sizeof(kHmacRefData) * 8);
+
+  // Load data into KMAC block.
+  CHECK_DIF_OK(dif_kmac_absorb(&kmac, &kmac_operation_state, kKmacMessage,
+                               kKmacMessageLength, NULL));
+
+  OTTF_TASK_DELETE_SELF_OR_DIE;
+}
+
+static void comms_data_load_task(void *task_parameters) {
+  LOG_INFO("Loading communication block FIFOs with data ...");
+  size_t bytes_written;
+  CHECK(ARRAYSIZE(kUartMessage) == 32);
+  dif_uart_t uart_handles[] = {uart_1, uart_2, uart_3};
+
+  // Load data into UART FIFO.
+  for (size_t i = 0; i < ARRAYSIZE(uart_handles); ++i) {
+    bytes_written = 0;
+    CHECK_DIF_OK(dif_uart_bytes_send(&uart_handles[i], kUartMessage,
+                                     ARRAYSIZE(kUartMessage), &bytes_written));
+    CHECK(bytes_written == ARRAYSIZE(kUartMessage));
+  }
+
+  OTTF_TASK_DELETE_SELF_OR_DIE;
+}
+
+static void max_power_task(void *task_parameters) {
+  LOG_INFO("Starting the max power task ...");
+  // ***************************************************************************
+  // Trigger all chip operations.
+  //
+  // Note: We trigger the activations of each operation manually, rather than
+  // use the DIFs, so that we can maximize the time overlap between all
+  // operations.
+  // ***************************************************************************
+
+  // Prepare KMAC for squeeze command. Note, below code is derived from the KMAC
+  // DIF `dif_kmac_sqeeze()`.
+  CHECK(!kmac_operation_state.squeezing);
+  if (kmac_operation_state.append_d) {
+    // The KMAC operation requires that the output length (d) in bits be right
+    // encoded and appended to the end of the message.
+    // Note: kDifKmacMaxOutputLenWords could be reduced to make this code
+    // simpler. For example, a maximum of `(UINT16_MAX - 32) / 32` (just under
+    // 8 KiB) would mean that d is guaranteed to be less than 0xFFFF.
+    uint32_t d = kmac_operation_state.d * 32;
+    int len = 1 + (d > 0xFF) + (d > 0xFFFF) + (d > 0xFFFFFF);
+    int shift = (len - 1) * 8;
+    while (shift >= 8) {
+      mmio_region_write8(kmac.base_addr, KMAC_MSG_FIFO_REG_OFFSET,
+                         (uint8_t)(d >> shift));
+      shift -= 8;
+    }
+    mmio_region_write8(kmac.base_addr, KMAC_MSG_FIFO_REG_OFFSET, (uint8_t)d);
+    mmio_region_write8(kmac.base_addr, KMAC_MSG_FIFO_REG_OFFSET, (uint8_t)len);
+  }
+
+  // Prepare AES and HMAC trigger / process commands.
+  uint32_t aes_trigger_reg = bitfield_bit32_write(0, kDifAesTriggerStart, true);
+  uint32_t hmac_cmd_reg =
+      mmio_region_read32(hmac.base_addr, HMAC_CMD_REG_OFFSET);
+  hmac_cmd_reg =
+      bitfield_bit32_write(hmac_cmd_reg, HMAC_CMD_HASH_PROCESS_BIT, true);
+  uint32_t kmac_cmd_reg =
+      bitfield_field32_write(0, KMAC_CMD_CMD_FIELD, KMAC_CMD_CMD_VALUE_PROCESS);
```

Comment:
KMAC also supports a `run` command to generate more state for non-fixed-length operations (SHAKE, cSHAKE, and KMAC, but not SHA3). It can be called an unlimited number of times, so if you want KMAC to run for a longer period of time it might be helpful to listen to the status reg and repeatedly issue `run` commands.

Reply:
Oh good to know. I was going to start with just a test that executes one "max power" epoch/period, but if we have to loop over several this will be helpful, thanks!